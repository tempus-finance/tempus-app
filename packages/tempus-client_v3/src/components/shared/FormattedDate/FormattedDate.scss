.tc__formatted-date {
  display: inline-flex;
  align-items: center;
}

.tc__formatted-date__separator {
  width: 4px;
  height: 4px;
  margin: 0 4px;
  border-radius: 50%;
}

.tc__formatted-date__separator-low-contrast {
  background-color: var(--formattedDateSeparatorLowContrast);
}

.tc__formatted-date__separator-high-contrast {
  background-color: var(--formattedDateSeparatorHighContrast);
}
