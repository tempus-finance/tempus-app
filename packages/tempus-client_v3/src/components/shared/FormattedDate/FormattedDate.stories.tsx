import { ComponentStory, ComponentMeta } from '@storybook/react';
import { CSSProperties } from 'react';
import FormattedDate from './FormattedDate';

export default {
  title: 'FormattedDate',
  component: FormattedDate,
  argTypes: {
    date: {
      control: { type: 'date' },
    },
    separatorContrast: {
      options: ['low', 'high'],
      control: { type: 'radio' },
    },
  },
} as ComponentMeta<typeof FormattedDate>;

const style: CSSProperties = {
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'center',
  justifyContent: 'center',
  padding: '10px',
};

const Template: ComponentStory<typeof FormattedDate> = props => {
  const { date, separatorContrast } = props;
  return (
    <div style={style}>
      <FormattedDate date={date} size="small" separatorContrast={separatorContrast} />
      <FormattedDate date={date} size="medium" separatorContrast={separatorContrast} />
      <FormattedDate date={date} size="large" separatorContrast={separatorContrast} />
    </div>
  );
};

export const StyledFormattedDate = Template.bind({});
StyledFormattedDate.args = {
  date: new Date(),
  separatorContrast: 'high',
};
