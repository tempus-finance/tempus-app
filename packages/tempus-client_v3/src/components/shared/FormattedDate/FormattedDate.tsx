import { FC, Fragment, memo, useMemo } from 'react';
import Typography, { TypographyColor, TypographyVariant } from '../Typography';
import './FormattedDate.scss';

type FormattedDateSize = 'small' | 'medium' | 'large';

export interface FormattedDateProps {
  date: Date;
  locale: string;
  textColor?: TypographyColor;
  size?: FormattedDateSize;
  separatorContrast?: 'low' | 'high';
  dateParts?: Set<Intl.DateTimeFormatPartTypes>;
}

const formattedDateTypographyVariantMap: Record<FormattedDateSize, TypographyVariant> = {
  small: 'body-secondary',
  medium: 'body-primary',
  large: 'subheader',
};

const formatOptions: Intl.DateTimeFormatOptions = {
  day: '2-digit',
  month: 'short',
  year: '2-digit',
};

const FormattedDate: FC<FormattedDateProps> = props => {
  const {
    date,
    locale,
    textColor,
    size = 'medium',
    separatorContrast = 'high',
    dateParts = new Set(['day', 'month', 'year']),
  } = props;
  const formattedDateParts = useMemo(
    () =>
      new Intl.DateTimeFormat(locale, formatOptions)
        .formatToParts(date)
        .filter(part => dateParts.has(part.type))
        .map(part => part.value.toString()),
    [date, locale, dateParts],
  );

  return (
    <span className="tc__formatted-date">
      {formattedDateParts.map((part, index) => (
        <Fragment key={`formatted-date-part-${index}`}>
          <Typography variant={formattedDateTypographyVariantMap[size]} type="mono" weight="medium" color={textColor}>
            {part}
          </Typography>
          {index < formattedDateParts.length - 1 && (
            <span
              className={`tc__formatted-date__separator tc__formatted-date__separator-${separatorContrast}-contrast`}
            />
          )}
        </Fragment>
      ))}
    </span>
  );
};

export default memo(FormattedDate);
