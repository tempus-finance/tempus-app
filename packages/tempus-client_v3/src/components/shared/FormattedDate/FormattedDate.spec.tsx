import { render } from '@testing-library/react';
import FormattedDate, { FormattedDateProps } from './FormattedDate';

const defaultProps: FormattedDateProps = {
  date: new Date(Date.UTC(2022, 1, 4)),
  locale: 'en-GB',
};

const subject = (props: FormattedDateProps) => render(<FormattedDate {...props} />);

describe('FormattedDate', () => {
  it('formats a Date like `04 - Feb - 22`', () => {
    const { queryByText, container } = subject(defaultProps);

    const actualFormattedDate = container.querySelector('.tc__formatted-date');
    const actualDayText = queryByText('04');
    const actualMonthText = queryByText('Feb');
    const actualYearText = queryByText('22');

    expect(actualYearText).not.toBeNull();
    expect(actualMonthText).not.toBeNull();
    expect(actualDayText).not.toBeNull();

    expect(actualFormattedDate).toMatchSnapshot();
  });
});
