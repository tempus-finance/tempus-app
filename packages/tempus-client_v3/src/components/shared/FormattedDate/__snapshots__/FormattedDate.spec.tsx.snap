// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormattedDate formats a Date like \`04 - Feb - 22\` 1`] = `
<span
  class="tc__formatted-date"
>
  <div
    style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 500; font-family: 'Azeret Mono', monospace; opacity: 1;"
  >
    04
  </div>
  <span
    class="tc__formatted-date__separator tc__formatted-date__separator-high-contrast"
  />
  <div
    style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 500; font-family: 'Azeret Mono', monospace; opacity: 1;"
  >
    Feb
  </div>
  <span
    class="tc__formatted-date__separator tc__formatted-date__separator-high-contrast"
  />
  <div
    style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 500; font-family: 'Azeret Mono', monospace; opacity: 1;"
  >
    22
  </div>
</span>
`;
